src/v/raft/replicate_batcher.cc

```diff
@@ -28,53 +32,60 @@ replicate_batcher::replicate_batcher(consensus* ptr, size_t cache_size)
 replicate_stages replicate_batcher::replicate(
   std::optional<model::term_id> expected_term,
   model::record_batch_reader r,
-  consistency_level consistency_lvl) {
+  consistency_level consistency_lvl,
+  std::optional<std::chrono::milliseconds> timeout) {
     ss::promise<> enqueued;
     auto enqueued_f = enqueued.get_future();
-    try {
-        auto f = cache_and_wait_for_result(
-          std::move(enqueued), expected_term, std::move(r), consistency_lvl);
-        return {std::move(enqueued_f), std::move(f)};
-    } catch (...) {
-        return {
-          ss::current_exception_as_future<>(),
-          ss::make_ready_future<result<replicate_result>>(errc::shutting_down)};
-    }
+
+    auto f = cache_and_wait_for_result(
+      std::move(enqueued),
+      expected_term,
+      std::move(r),
+      consistency_lvl,
+      timeout);
+    return {std::move(enqueued_f), std::move(f)};
 }
 
 ss::future<result<replicate_result>>
 replicate_batcher::cache_and_wait_for_result(
   ss::promise<> enqueued,
   std::optional<model::term_id> expected_term,
   model::record_batch_reader r,
-  consistency_level consistency_lvl) {
+  consistency_level consistency_lvl,
+  std::optional<std::chrono::milliseconds> timeout) {
+    item_ptr item;
     try {
         auto holder = _bg.hold();
-        auto item = co_await do_cache(
-          expected_term, std::move(r), consistency_lvl);
+        item = co_await do_cache(
+          expected_term, std::move(r), consistency_lvl, timeout);
+
         // now request is already enqueued, we can release first
         // stage future
         enqueued.set_value();
-        try {
-            auto u = co_await _lock.get_units();
-            co_await flush(std::move(u), false);
-            _ptr->_probe.replicate_done();
-            // release holder before waiting for the future
-            holder.release();
```

Comment:
i assume that it was intentional that you wanted to keep the gate held longer (since releasing it here was dropped)?

Reply:
yes, it wouldn't be needed as now the `item-get_future()` is called outside of the try/catch block so the gate is released before